File: 0046. Permutations/Permutations.java
class Solution {
    private void backtracking(int[] nums, List<Integer> tmp, List<List<Integer>> res) {
        if (tmp.size() == nums.length) {
            res.add(new ArrayList<>(tmp));
        } else {
            for (int num : nums) {
                if (tmp.contains(num)) continue;
                tmp.add(num);
                backtracking(nums, tmp, res);
                tmp.remove(tmp.size() - 1);
            }
        }
    }
    
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        backtracking(nums, new ArrayList<>(), res);
        return res;
    }
}

File: 0046. Permutations/Permutations.cpp
class Solution {
    void backtracking(vector<int>& nums, vector<int> &tmp, vector<vector<int>> &res) {
        if (tmp.size() == nums.size()) {
            res.push_back(tmp);
        } else {
            for (int i = 0; i < nums.size(); i++) {
                if (find(tmp.begin(), tmp.end(), nums[i]) != tmp.end()) continue;
                tmp.push_back(nums[i]);
                backtracking(nums, tmp, res);
                tmp.pop_back();
            }
        }
    } 
        
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int> tmp;
        vector<vector<int>> res;
        backtracking(nums, tmp, res);
        return res;
    }
};
